File: tests/perl/src/T001_Status.pl
#!/usr/bin/env perl

use strict;
use Test;
use libgmsec_perl;


sub Test_Status
{
	my ($test) = @_;

	$test->setDescription("Test Status");

	test_constructor_1($test);
	test_constructor_2($test);
	test_exception_constructor($test);
	test_copy_constructor($test);
	test_get($test);
	test_set($test);
	test_set_class($test);
	test_set_code($test);
	test_set_reason($test);
	test_set_custom_code($test);
	test_reset($test);
}


sub test_constructor_1
{
	libgmsec_perl::logInfo("Test constructor (no args)");

	my ($test) = @_;

	my $reason = "";

   	my $s1 = libgmsec_perl::Status->new();

	$test->check("Unexpected error", $s1->hasError() == 0);
	$test->check("Unexpected classification", $s1->getClass() == $libgmsec_perl::NO_ERROR_CLASS);
	$test->check("Unexpected code", $s1->getCode() == $libgmsec_perl::NO_ERROR_CODE);
	$test->check("Unexpected custom code", $s1->getCustomCode() == 0);
	$test->check("Unexpected reason", $s1->getReason() eq $reason);
}


sub test_constructor_2
{
	libgmsec_perl::logInfo("Test constructor (2-4 args)");

	my ($test) = @_;

	my $reason = "Some worthy status";

	my $s1 = libgmsec_perl::Status->new(1, 2);

	$test->check("Unexpected error", $s1->hasError());
	$test->check("Unexpected classification", $s1->getClass() == 1);
	$test->check("Unexpected code", $s1->getCode() == 2);
	$test->check("Unexpected custom code", $s1->getCustomCode() == 0);
	$test->check("Unexpected reason", $s1->getReason() eq "");

	my $s2 = libgmsec_perl::Status->new(3, 4, $reason);

	$test->check("Unexpected error", $s2->hasError());
	$test->check("Unexpected classification", $s2->getClass() == 3);
	$test->check("Unexpected code", $s2->getCode() == 4);
	$test->check("Unexpected custom code", $s2->getCustomCode() == 0);
	$test->check("Unexpected reason", $s2->getReason() eq $reason);

	my $s3 = libgmsec_perl::Status->new(6, 7, $reason, 8);

	$test->check("Unexpected error", $s3->hasError());
	$test->check("Unexpected classification", $s3->getClass() == 6);
	$test->check("Unexpected code", $s3->getCode() == 7);
	$test->check("Unexpected custom code", $s3->getCustomCode() == 8);
	$test->check("Unexpected reason", $s3->getReason() eq $reason);
}


sub test_exception_constructor
{
	libgmsec_perl::logInfo("Test constructor (GmsecError arg)");

	my ($test) = @_;

   	my $reason = "Some worthy exception";

	my $e1 = libgmsec_perl::GmsecException->new(1, 2, 3, $reason);

	my $s1 = libgmsec_perl::Status->new($e1);

	$test->check("Unexpected error", $s1->hasError());
	$test->check("Unexpected classification", $s1->getClass() == 1);
	$test->check("Unexpected code", $s1->getCode() == 2);
	$test->check("Unexpected custom code", $s1->getCustomCode() == 3);
	$test->check("Unexpected reason", $s1->getReason() eq $reason);
}


sub test_copy_constructor
{
	libgmsec_perl::logInfo("Test copy-constructor");

	my ($test) = @_;

	my $reason = "Some worthy exception";

	my $s1 = libgmsec_perl::Status->new(1, 2, $reason, 3);
	my $s2 = libgmsec_perl::Status->new($s1);

	$test->check("Unexpected error", $s2->hasError());
	$test->check("Unexpected classification", $s1->getClass() == $s2->getClass());
	$test->check("Unexpected code", $s1->getCode() == $s2->getCode());
	$test->check("Unexpected custom code", $s1->getCustomCode() == $s2->getCustomCode());
	$test->check("Unexpected reason", $s2->getReason() eq $reason);
}


sub test_get
{
	libgmsec_perl::logInfo("Test get()");

	my ($test) = @_;

	my $reason = "Some worthy exception";

	my $s1 = libgmsec_perl::Status->new(1, 2, $reason, 3);

	$test->check("Unexpected get() string", $s1->get() == "[1,2,3] : " . $reason);
}


sub test_set
{
	libgmsec_perl::logInfo("Test set()");

	my ($test) = @_;

	my $reason = "Some worthy exception";

	my $s1 = libgmsec_perl::Status->new();
	$s1->set(1, 2, $reason);
	$test->check("Unexpected error", $s1->hasError());
	$test->check("Unexpected classification", $s1->getClass() == 1);
	$test->check("Unexpected code", $s1->getCode() == 2);
	$test->check("Unexpected custom code", $s1->getCustomCode() == 0);
	$test->check("Unexpected reason", $s1->getReason() eq $reason);

	my $s2 = libgmsec_perl::Status->new(1, 2, "foobar", 6);
	$s2->set(1, 2, $reason, 3);
	$test->check("Unexpected error", $s2->hasError());
	$test->check("Unexpected classification", $s2->getClass() == 1);
	$test->check("Unexpected code", $s2->getCode() == 2);
	$test->check("Unexpected custom code", $s2->getCustomCode() == 3);
	$test->check("Unexpected reason", $s2->getReason() eq $reason);
}


sub test_set_class
{
	libgmsec_perl::logInfo("Test set_class()");

	my ($test) = @_;

	my $s1 = libgmsec_perl::Status->new();
	$s1->setClass(0);
	$test->check("Unexpected error", $s1->hasError() == 0);
	$test->check("Unexpected classification", $s1->getClass() == 0);
	$test->check("Unexpected code", $s1->getCode() == 0);
	$test->check("Unexpected custom code", $s1->getCustomCode() == 0);
	$test->check("Unexpected reason", $s1->getReason() eq "");

	my $s2 = libgmsec_perl::Status->new();
	$s2->setClass(1);
	$test->check("Unexpected error", $s2->hasError());
	$test->check("Unexpected classification", $s2->getClass() == 1);
	$test->check("Unexpected code", $s2->getCode() == 0);
	$test->check("Unexpected custom code", $s2->getCustomCode() == 0);
	$test->check("Unexpected reason", $s2->getReason() eq "");
}


sub test_set_code
{
	libgmsec_perl::logInfo("Test set_code()");

	my ($test) = @_;

	my $s1 = libgmsec_perl::Status->new();
	$s1->setCode(1);
	$test->check("Unexpected error", $s1->hasError() == 0);
	$test->check("Unexpected classification", $s1->getClass() == 0);
	$test->check("Unexpected code", $s1->getCode() == 1);
	$test->check("Unexpected custom code", $s1->getCustomCode() == 0);
	$test->check("Unexpected reason", $s1->getReason() eq "");
}


sub test_set_reason
{
	libgmsec_perl::logInfo("Test set_reason()");

	my ($test) = @_;

	my $reason = "Some worthy exception";

	my $s1 = libgmsec_perl::Status->new();
	$s1->setReason($reason);
	$test->check("Unexpected error", $s1->hasError() == 0);
	$test->check("Unexpected classification", $s1->getClass() == 0);
	$test->check("Unexpected code", $s1->getCode() == 0);
	$test->check("Unexpected custom code", $s1->getCustomCode() == 0);
	$test->check("Unexpected reason", $s1->getReason() eq $reason);
}


sub test_set_custom_code
{
	libgmsec_perl::logInfo("Test set_custom_code()");

	my ($test) = @_;

	my $s1 = libgmsec_perl::Status->new();
	$s1->setCustomCode(1);
	$test->check("Unexpected error", $s1->hasError() == 0);
	$test->check("Unexpected classification", $s1->getClass() == 0);
	$test->check("Unexpected code", $s1->getCode() == 0);
	$test->check("Unexpected custom code", $s1->getCustomCode() == 1);
	$test->check("Unexpected reason", $s1->getReason() eq "");
}


sub test_reset
{
	libgmsec_perl::logInfo("Test reset()");

	my ($test) = @_;

	my $s1 = libgmsec_perl::Status->new(1, 2, "foobar", 3);
	$s1->reset();
	$test->check("Unexpected error", $s1->hasError() == 0);
	$test->check("Unexpected classification", $s1->getClass() == 0);
	$test->check("Unexpected code", $s1->getCode() == 0);
	$test->check("Unexpected custom code", $s1->getCustomCode() == 0);
	$test->check("Unexpected reason", $s1->getReason() eq "");
}


Test::driver(\&Test_Status);

File: tests/nodejs/src/T001_Status.js
#!/usr/bin/env node

const TestCase = require('./TestCase.js');
const gmsec = require('../../../bin/gmsec_nodejs.node');


class T001_Status extends TestCase
{
	constructor()
	{
		super();
	}

	testCase()
	{
		this.testConstructor_1();
		this.testConstructor_2();
		this.testCopyConstructor();
		this.testGet();
		this.testSet();
		this.testSetClass();
		this.testSetCode();
		this.testSetReason();
		this.testSetCustomCode();
		this.testReset();
	}

	testConstructor_1()
	{
		gmsec.Log.info("Test constructor (no args)");

		var s1 = new gmsec.Status();
		this.check("Unexpected error", s1.hasError() == false);
		this.check("Unexpected class", s1.getClass() == gmsec.NO_ERROR_CLASS);
		this.check("Unexpected code", s1.getCode() == gmsec.NO_ERROR_CODE);
		this.check("Unexpected custom code", s1.getCustomCode() == 0);
		this.check("Unexpected reason", s1.getReason() == "");
	}

	testConstructor_2()
	{
		gmsec.Log.info("Test constructor (2-4 args)");

		var s1 = new gmsec.Status(1, 2);
		this.check("Unexpected error", s1.hasError());
		this.check("Unexpected class", s1.getClass() == 1);
		this.check("Unexpected code", s1.getCode() == 2);
		this.check("Unexpected custom code", s1.getCustomCode() == 0);
		this.check("Unexpected reason", s1.getReason() == "");

		var s2 = new gmsec.Status(3, 4, "Some worthy status");
		this.check("Unexpected error", s2.hasError());
		this.check("Unexpected class", s2.getClass() == 3);
		this.check("Unexpected code", s2.getCode() == 4);
		this.check("Unexpected custom code", s2.getCustomCode() == 0);
		this.check("Unexpected reason", s2.getReason() == "Some worthy status");

		var s3 = new gmsec.Status(5, 6, "Some worthy status", 7);
		this.check("Unexpected error", s3.hasError());
		this.check("Unexpected class", s3.getClass() == 5);
		this.check("Unexpected code", s3.getCode() == 6);
		this.check("Unexpected custom code", s3.getCustomCode() == 7);
		this.check("Unexpected reason", s3.getReason() == "Some worthy status");
	}

	testCopyConstructor()
	{
		gmsec.Log.info("Test copy constructor");

		var s1 = new gmsec.Status(1, 2, "Some worthy status", 3);
		var s2 = new gmsec.Status(s1);

		this.check("Unexpected error", s1.hasError() == s2.hasError());
		this.check("Unexpected class", s1.getClass() == s2.getClass());
		this.check("Unexpected code", s1.getCode() == s2.getCode());
		this.check("Unexpected custom code", s1.getCustomCode() == s2.getCustomCode());
		this.check("Unexpected reason", s1.getReason() == s2.getReason());
	}

	testGet()
	{
		gmsec.Log.info("Test get()");

		var s1 = new gmsec.Status(1, 2, "Some worthy status", 3);

		this.check("Unexpected get string", s1.get() == "[1,2,3] : Some worthy status");
	}

	testSet()
	{
		gmsec.Log.info("Test set()");

		var s1 = new gmsec.Status();
		s1.set(0, 1);
		this.check("Unexpected error", s1.hasError() == false);
		this.check("Unexpected class", s1.getClass() == 0);
		this.check("Unexpected code", s1.getCode() == 1);
		this.check("Unexpected custom code", s1.getCustomCode() == 0);
		this.check("Unexpected reason", s1.getReason() == "");

		var s2 = new gmsec.Status();
		s2.set(1, 2, "Some worthy status");
		this.check("Unexpected error", s2.hasError());
		this.check("Unexpected class", s2.getClass() == 1);
		this.check("Unexpected code", s2.getCode() == 2);
		this.check("Unexpected custom code", s2.getCustomCode() == 0);
		this.check("Unexpected reason", s2.getReason() == "Some worthy status");

		var s3 = new gmsec.Status(1, 2, "foobar", 6);
		s3.set(1, 2, "Some worthy status", 3);
		this.check("Unexpected error", s3.hasError());
		this.check("Unexpected class", s3.getClass() == 1);
		this.check("Unexpected code", s3.getCode() == 2);
		this.check("Unexpected custom code", s3.getCustomCode() == 3);
		this.check("Unexpected reason", s3.getReason() == "Some worthy status");
	}

	testSetClass()
	{
		gmsec.Log.info("Test setClass()");

		var s1 = new gmsec.Status();
		s1.setClass(0);
		this.check("Unexpected error", s1.hasError() == false);
		this.check("Unexpected class", s1.getClass() == 0);
		this.check("Unexpected code", s1.getCode() == 0);
		this.check("Unexpected custom code", s1.getCustomCode() == 0);
		this.check("Unexpected reason", s1.getReason() == "");

		var s2 = new gmsec.Status();
		s2.setClass(1);
		this.check("Unexpected error", s2.hasError());
		this.check("Unexpected class", s2.getClass() == 1);
		this.check("Unexpected code", s2.getCode() == 0);
		this.check("Unexpected custom code", s2.getCustomCode() == 0);
		this.check("Unexpected reason", s2.getReason() == "");
	}

	testSetCode()
	{
		gmsec.Log.info("Test setCode()");

		var s1 = new gmsec.Status();
		s1.setCode(1);
		this.check("Unexpected error", s1.hasError() == false);
		this.check("Unexpected class", s1.getClass() == 0);
		this.check("Unexpected code", s1.getCode() == 1);
		this.check("Unexpected custom code", s1.getCustomCode() == 0);
		this.check("Unexpected reason", s1.getReason() == "");
	}

	testSetReason()
	{
		gmsec.Log.info("Test setReason()");

		var s1 = new gmsec.Status();
		s1.setReason("Some worthy status");
		this.check("Unexpected error", s1.hasError() == false);
		this.check("Unexpected class", s1.getClass() == 0);
		this.check("Unexpected code", s1.getCode() == 0);
		this.check("Unexpected custom code", s1.getCustomCode() == 0);
		this.check("Unexpected reason", s1.getReason() == "Some worthy status");
	}

	testSetCustomCode()
	{
		gmsec.Log.info("Test setCustomCode()");

		var s1 = new gmsec.Status();
		s1.setCustomCode(1);
		this.check("Unexpected error", s1.hasError() == false);
		this.check("Unexpected class", s1.getClass() == 0);
		this.check("Unexpected code", s1.getCode() == 0);
		this.check("Unexpected custom code", s1.getCustomCode() == 1);
		this.check("Unexpected reason", s1.getReason() == "");
	}

	testReset()
	{
		gmsec.Log.info("Test reset()");

		var s1 = new gmsec.Status(1, 2, "foobar", 3);
		s1.reset();
		this.check("Unexpected error", s1.hasError() == false);
		this.check("Unexpected class", s1.getClass() == 0);
		this.check("Unexpected code", s1.getCode() == 0);
		this.check("Unexpected custom code", s1.getCustomCode() == 0);
		this.check("Unexpected reason", s1.getReason() == "");
	}
}

test = new T001_Status();
test.setDescription("Test Status");
test.run(process.argv);
